street_network/src/transform/dual_carriageways.rs

```diff
@@ -1,55 +1,350 @@
-use std::collections::BTreeSet;
+use std::collections::{BTreeMap, BTreeSet};
+
+use geom::Distance;
 
 use crate::{osm, OriginalRoad, StreetNetwork};
 
-/// Does this road go between two divided one-ways? Ideally they're tagged explicitly
-/// (https://wiki.openstreetmap.org/wiki/Tag:dual_carriageway%3Dyes), but we can also apply simple
-/// heuristics to guess this.
-#[allow(unused)]
-pub fn connects_dual_carriageway(streets: &StreetNetwork, id: &OriginalRoad) -> bool {
-    let connectors_angle = streets.roads[id].angle();
-    // There are false positives like https://www.openstreetmap.org/way/4636259 when we're looking
-    // at a segment along a marked dual carriageway. Filter out by requiring the intersecting dual
-    // carriageways to differ by a minimum angle.
-    let within_degrees = 10.0;
-
-    let mut i1_dual_carriageway = false;
-    let mut oneway_names_i1: BTreeSet<String> = BTreeSet::new();
-    for r in streets.roads_per_intersection(id.i1) {
-        let road = &streets.roads[&r];
-        if r == *id || connectors_angle.approx_eq(road.angle(), within_degrees) {
-            continue;
-        }
-        if road.osm_tags.is("dual_carriageway", "yes") {
-            i1_dual_carriageway = true;
-        }
-        if road.oneway_for_driving().is_some() {
+pub fn merge(streets: &mut StreetNetwork) {
+    for i in streets.intersections.keys() {
+        // Progressively detect more stuff. Display the most detail possible.
+        if let Some(mc) = MultiConnection::new(streets, *i) {
+            // TODO Ignore opposite direction of one we've already found?
+            if let Some(dc1) = DualCarriagewayPt1::new(streets, &mc) {
+                if let Some(dc2) = DualCarriagewayPt2::new(streets, &dc1) {
+                    dc2.debug(streets);
+                } else {
+                    dc1.debug(streets);
+                }
+            } else {
+                mc.debug(streets);
+            }
+
+            // TODO Just work on one right now
+            break;
+        }
+    }
+}
+
+// TODO We should do this in classify_intersections.rs?
```

Comment:
Maybe this should happen in `classify_intersections.rs` upfront. I'm not trying to reason about highway entrances/exits with this transformation, so maybe the pattern matching doesn't apply completely